feat(components): add `ui` field in items

J-Michalek opened the pull request:
### 🔗 Linked issue

Resolves: https://github.com/nuxt/ui/issues/4059

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

<!-- Describe your changes in detail -->
I have added the `ui` field in `items` array - this affects all components that currently use an `items` prop.

<!-- Why is this change required? What problem does it solve? -->
This will simplify cases where the user wants to use a specific classes for different parts of an item but does not want to affect other items at the same time.

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

J-Michalek committed:
feat(stepper): add iconClass prop to stepper items

J-Michalek committed:
test(stepper): add case for iconClass prop

J-Michalek committed:
docs(stepper): add iconClass prop

J-Michalek committed:
test(stepper): remove obsolete snapshot entry

J-Michalek committed:
test(stepper): update snapshot for vue tests

benjamincanac commented:
I agree this is a nice addition! However, this change should be applied to all components with an `items` prop for consistency 😬

J-Michalek commented:
@benjamincanac I can work on this. Should I bunch it up into a one big PR or should I create separate PR for each instance?

benjamincanac commented:
Great! No you can do this in this PR 😊

J-Michalek commented:
Quick checklist to track the progress on this (I will validate it as I go through it):
- [x] Accordion
- [x] Tabs
- [x] Stepper
- [x] Carousel
- [x] Breadcrumb
- [x] CheckboxGroup
- [x] RadioGroup
- [x] InputMenu
- [x] Tree
- [x] Select
- [x] SelectMenu
- [x] NavigationMenu - here we could support different ui fields for `NavigationMenuItem` and `NavigationMenuChildItem` but that would required some refactoring of the component, perhaps in another iteration.
- [x] DropdownMenu
- [x] ContextMenu
- [x] CommandPallete

J-Michalek committed:
Merge branch 'nuxt:v3' into feat/stepper-icon-class

J-Michalek commented:
@benjamincanac Just one more idea, what if we added a `ui` prop to the item interface that would allow customization of each part of the item but on a per item basis.

i.e. for the Accordion each item could have a `ui` prop that would contain the slots that affect the item specifically:
```vue
<AccordionItem
      v-for="(item, index) in props.items"
      v-slot="{ open }"
      :key="index"
      :value="item.value || String(index)"
      :disabled="item.disabled"
      :class="ui.item({ class: [props.ui?.item, item.ui?.item] })"
    >
      <AccordionHeader as="div" :class="ui.header({ class: [props.ui?.header, item.ui?.header] })">
        <AccordionTrigger :class="ui.trigger({ class: [props.ui?.trigger, item.ui?.trigger], disabled: item.disabled })">
          <slot name="leading" :item="item" :index="index" :open="open">
            <UIcon v-if="item.icon" :name="item.icon" :class="ui.leadingIcon({ class: [props.ui?.leadingIcon, item?.ui?.leadingIcon] })" />
          </slot>

          <span v-if="get(item, props.labelKey as string) || !!slots.default" :class="ui.label({ class: [props.ui?.label, item.ui?.label] })">
            <slot :item="item" :index="index" :open="open">{{ get(item, props.labelKey as string) }}</slot>
          </span>

          <slot name="trailing" :item="item" :index="index" :open="open">
            <UIcon :name="item.trailingIcon || trailingIcon || appConfig.ui.icons.chevronDown" :class="ui.trailingIcon({ class: [props.ui?.trailingIcon, item.ui?.trailingIcon] })" />
          </slot>
        </AccordionTrigger>
      </AccordionHeader>

      <AccordionContent v-if="item.content || !!slots.content || (item.slot && !!slots[item.slot as keyof AccordionSlots<T>]) || !!slots.body || (item.slot && !!slots[`${item.slot}-body` as keyof AccordionSlots<T>])" :class="ui.content({ class: [props.ui?.content, item.ui?.content] })">
        <slot :name="((item.slot || 'content') as keyof AccordionSlots<T>)" :item="(item as Extract<T, { slot: string; }>)" :index="index" :open="open">
          <div :class="ui.body({ class: [props.ui?.body, item.ui?.body] })">
            <slot :name="((item.slot ? `${item.slot}-body`: 'body') as keyof AccordionSlots<T>)" :item="(item as Extract<T, { slot: string; }>)" :index="index" :open="open">
              {{ item.content }}
            </slot>
          </div>
        </slot>
      </AccordionContent>
    </AccordionItem>
```

I feel like it would be more future proof as everyone might have different needs of per item customization, of course this is more involved solution but I think it might be worth it.

benjamincanac commented:
We could uniformize this as well indeed, I did it in very specific places where I needed to make an example only at the moment: https://github.com/nuxt/ui/blob/v3/src/runtime/components/Select.vue#L243

J-Michalek committed:
feat(accordion): add ui field in items

J-Michalek committed:
Merge branch 'feat/stepper-icon-class' of https://github.com/J-Michalek/ui into feat/stepper-icon-class

J-Michalek commented:
@benjamincanac Here is a quick [commit ](https://github.com/nuxt/ui/pull/4060/commits/1197806d6a6ba13be3a55c96746fd3bc10a3e2ea)for the accordion component, if you could quickly have a look if I'm going in the right direction with this.

J-Michalek committed:
test(accordion): add new snapshot entry for vue tests

J-Michalek committed:
docs(accordion): add ui field in items

Maybe I'm being too specific with this? The use can see it in detail in the props section.

J-Michalek committed:
feat(tabs): add ui field on items

J-Michalek committed:
docs(tabs): add ui field on items

J-Michalek committed:
test(tabs): add ui field on items

Commit:
feat(stepper): add ui field on items

Commit:
test(stepper): add ui field on items

Commit:
docs(stepper): add ui field on items

Commit:
feat(breadcrumb): add ui field on items

Commit:
test(breadcrumb): add ui field on items

Commit:
docs(breadcrumb): add ui field on items

Commit:
feat(checkbox-group): add ui field on items

Commit:
test(checkbox-group): add ui field on items

Commit:
docs(checkbox-group): add ui field on items

Commit:
docs(breadcrumb): fix description for object type

benjamincanac commented:
That is not a bad idea to handle this with a `ui` object in the items actually, do we still need `iconClass` though?

J-Michalek commented:
@benjamincanac No, we definitely don't need the `iconClass` field in that case. Also with this approach we could remove the [item.class](https://github.com/nuxt/ui/blob/v3/src/runtime/components/Breadcrumb.vue#L89) bindings that are sometimes used, because the `ui` field could be used instead - this would be a breaking change so definitely not in the scope of this PR.

Commit:
feat(radio-group): add ui field on items

Commit:
test(radio-group): add ui field on items

Commit:
docs(radio-group): add ui field on items

Commit:
feat(input-menu): add ui field on items

Commit:
test(input-menu): add ui field on items

Commit:
docs(input-menu): add ui field on items

Commit:
feat(tree): add ui field on items

Commit:
test(tree): add ui field on items

Commit:
docs(tree): add ui field on items

J-Michalek committed:
Merge branch 'nuxt:v3' into feat/stepper-icon-class

J-Michalek committed:
feat(select): add ui field in items

J-Michalek committed:
test(select): add ui field in items

J-Michalek committed:
docs(select): add ui field in items

J-Michalek committed:
test(according): unite title for ui field in items cases

J-Michalek committed:
test(tabs): unite title for ui field in items cases

J-Michalek committed:
feat(select-menu): add ui field on items

J-Michalek committed:
test(select-menu): add ui field on items

J-Michalek committed:
docs(select-menu): add ui field on items

J-Michalek committed:
feat(navigation-menu): add ui field on items

J-Michalek committed:
test(navigation-menu): add ui field on items

J-Michalek committed:
docs(navigation-menu): add ui field on items

J-Michalek committed:
feat(dropdown-menu): add ui field on items

J-Michalek committed:
test(dropdown-menu): add ui field on items

J-Michalek committed:
docs(dropdown-menu): add ui field on items

J-Michalek committed:
fix(dropdown-menu): add missing label ui field on items

J-Michalek committed:
fix(dropdown-menu): add missing separator ui field on items

J-Michalek committed:
feat(context-menu): add ui field on items

J-Michalek committed:
test(context-menu): add ui field on items

J-Michalek committed:
docs(context-menu): add ui field on items

J-Michalek committed:
feat(carousel): add ui field on items

J-Michalek committed:
test(carousel): add ui field on items

J-Michalek committed:
docs(carousel): add ui field on items

J-Michalek committed:
test(carousel): fix listing issues

J-Michalek commented:
@benjamincanac I think I'm done with this. I have updated the PR description above.

Two things worth mentioning:
- I had to change the type of carousel items a bit, see https://github.com/nuxt/ui/pull/4060/commits/2a270eaa0354bf1c843d5cf26c561b89e375b600
- `NavigationMenu` is a bit more complicated because it has another array of nested items which would need different type for the `ui` field and we would need to differentiate between top level and child items in the template, for now I just implemented the ui field for the top level items and omitted it from the `NavigationMenuChildItem` interface, see https://github.com/nuxt/ui/pull/4060/commits/1214fcb2433f5c8cd47087e5d61b9d41790b3b8b#diff-c7d773da155e1f06395d05bef2ab7ad26a66ca1af0de932ddf88350acfc16a08R11

J-Michalek committed:
Merge branch 'v3' into feat/stepper-icon-class

benjamincanac commented on docs/content/3.components/stepper.md:
```diff
@@ -20,9 +20,11 @@ Use the `items` prop as an array of objects with the following properties:
 - `description?: AvatarProps`{lang="ts-type"}
 - `content?: string`{lang="ts-type"}
 - `icon?: string`{lang="ts-type"}
+- `iconClass?: string`{lang="ts-type"}
```
This should be removed.

benjamincanac commented on docs/content/3.components/carousel.md:
```diff
@@ -16,6 +16,9 @@ links:
 
 Use the `items` prop as an array and render each item using the default slot:
 
+You can also pass an array of objects with the following properties:
```
This should happen after the example for consistency with other components.

benjamincanac commented on src/runtime/components/Carousel.vue:
```diff
@@ -288,7 +297,7 @@ defineExpose({
           role="group"
           aria-roledescription="slide"
-          :class="ui.item({ class: props.ui?.item })"
+          :class="ui.item({ class: [props.ui?.item, isCarouselItem(item) && item.ui?.item] })"
```
I wonder if `item.class` would be easier here 🤔

J-Michalek replied:
@benjamincanac I suppose this is the question of consistency vs convenience. For consistency `ui` field would be suitable and would allow for extension if for whatever reason more slots were to be added (not likely I guess). For convenience of the developer `item.class` makes sense.

benjamincanac replied:
I think we can support both like in actual components, `class` would be for the "root" item and `ui` for what's inside. wdyt?

J-Michalek replied:
I could see that. Just to be clear would the `ui` field also include the "root" slot or would it be only accessible through `item.class`? I think we should add it to both, because if the `ui` has more fields I would rather just extend the `ui` and add the "root" to it, but if I only want to change the "root"  then `item.class` seems easier to use.

I would need to make a revision of the targeted components to make sure all of them allow to pass in `class` for the item.

benjamincanac commented on src/runtime/components/CheckboxGroup.vue:
```diff
@@ -14,6 +14,7 @@ export type CheckboxGroupItem = {
   description?: string
   disabled?: boolean
   value?: string
+  ui?: Pick<CheckboxGroup['slots'], 'item'>
```
I wonder if `item.class` would be easier here 🤔

J-Michalek committed:
docs(stepper): remove iconClass field

J-Michalek committed:
docs(carousel): adjust objects description

J-Michalek committed:
Merge branch 'v3' into feat/stepper-icon-class

J-Michalek committed:
Merge branch 'feat/stepper-icon-class' of https://github.com/J-Michalek/ui into feat/stepper-icon-class

Commit:
feat(components): add class field to items along with the ui field

Commit:
Merge branch 'v3' of https://github.com/J-Michalek/ui into feat/stepper-icon-class

Commit:
test(navigation-menu): update snapshots

J-Michalek commented:
I have added the `item.class` field to all of the affected components, see https://github.com/nuxt/ui/pull/4060/commits/229a04d99ce90af5c0a97274645a0eca82fe3878

I've encountered that the `Breadcrumb` and `NavigationMenu` components already use `item.class` and bind it to other elements that are not the item's root element 
- https://github.com/nuxt/ui/blob/v3/src/runtime/components/Breadcrumb.vue#L89
- https://github.com/nuxt/ui/blob/v3/src/runtime/components/NavigationMenu.vue#L267
we cannot easily change it as it would be a breaking change, but I think it is worth highlighting.

benjamincanac commented:
I guess you can change it, I'd rather introduce a small breaking change and have everything consistent. We'll release this in the next minor and explain it in the Breaking Changes section.

Commit:
feat(breadcrumb)!: move item.class to root element of the item

This is a small breaking change, but for consistency it was done.

Commit:
feat(navigation-menu)!: move item.class to root element of the item

This is a small breaking change, but for consistency it was done.

Commit:
Merge branch 'v3' of https://github.com/J-Michalek/ui into feat/stepper-icon-class

J-Michalek committed:
Merge branch 'v3' into feat/stepper-icon-class

J-Michalek commented:
Alright, I've updated it and made the small breaking changes.

https://github.com/nuxt/ui/pull/4060/commits/e9337e29b2945e7d09f9214ef99a2ad575b311ae
https://github.com/nuxt/ui/pull/4060/commits/72dfb4d6d800cfd8074df8ae36798e7ad7d8c760

The users affected by this breaking change will be able to use the new `ui` field on the item to get the same functionality.

benjamincanac commented:
Why doesn't it use `item.ui.link` instead of `item.class`? 🤔

Commit:
fix(navigation-menu): bind item.ui.link to ULinkBase

I overlooked this.

Commit:
Merge branch 'feat/stepper-icon-class' of https://github.com/J-Michalek/ui into feat/stepper-icon-class

J-Michalek commented:
My mistake, you got a sharp eyes! Thanks for your reviews btw.

Fixed in https://github.com/nuxt/ui/pull/4060/commits/feb7ca7379428c82c7a27cde5d65db6a7dc7db1e

benjamincanac committed:
Merge branch 'v3' into feat/stepper-icon-class

benjamincanac committed:
docs: move `class` under `ui`

benjamincanac committed:
fix: wrong overrides

benjamincanac committed:
test: update snapshots

benjamincanac committed:
chore: move `class` over `ui` for consistency

benjamincanac committed:
fix(CheckboxGroup): proxy checkbox ui

benjamincanac committed:
fix(ContextMenu/DropdownMenu): handle item.class for all item types

benjamincanac committed:
fix(InputMenu/Select/SelectMenu): handle item.class for all item types

benjamincanac committed:
docs(checkbox-group): missing ui

benjamincanac commented:
@J-Michalek I've pushed a few fixes, let me know what you think! Is there a reason you ommitted the CommandPalette component?

benjamincanac committed:
fix(Stepper): remove `iconClass`

benjamincanac commented on test/components/Accordion.spec.ts:
```diff
@@ -56,7 +56,26 @@ describe('Accordion', () => {
     ['with body slot', { props: { ...props, modelValue: '1' }, slots: { body: () => 'Body slot' } }],
     ['with custom slot', { props: { ...props, modelValue: '5' }, slots: { custom: () => 'Custom slot' } }],
-    ['with custom body slot', { props: { ...props, modelValue: '5' }, slots: { 'custom-body': () => 'Custom body slot' } }]
+    ['with custom body slot', { props: { ...props, modelValue: '5' }, slots: { 'custom-body': () => 'Custom body slot' } }],
```
I'm wondering if it's a good idea to test everything `item.ui` here, it's a lot and will be a pain to maintain 😅

J-Michalek replied:
Yeah I was unsure if I should add test cases for it, I guess we can remove it it is easily the largest block of test code for all of the affected component.

Commit:
test(accordion): remove duplicated tests

Commit:
feat(command-palette): add ui and class fields to items

benjamincanac commented on docs/content/3.components/command-palette.md:
```diff
@@ -34,6 +34,8 @@ The CommandPalette component filters groups and ranks matching commands by relev
 - [`ignoreFilter?: boolean`{lang="ts-type"}](#with-ignore-filter)
 - [`postFilter?: (searchTerm: string, items: T[]) => T[]`{lang="ts-type"}](#with-post-filtered-items)
 - `highlightedIcon?: string`{lang="ts-type"}
+- `ui?: { item?: ClassNameValue, itemLeadingIcon?: ClassNameValue, itemLeadingAvatarSize?: ClassNameValue, itemLeadingAvatar?: ClassNameValue, itemLeadingChipSize?: ClassNameValue, itemLeadingChip?: ClassNameValue, itemLabel?: ClassNameValue, itemLabelPrefix?: ClassNameValue, itemLabelBase?: ClassNameValue, itemLabelSuffix?: ClassNameValue, itemTrailing?: ClassNameValue, itemTrailingKbds?: ClassN ...
```
Shouldn't this be in `items` instead of `groups`? 🤔 I guess you can just remove it from the docs or enhance the `items` definition.

Commit:
test(components): remove cases for ui and class fields in items

It would be annoying to maintain these.

Commit:
docs(command-pallete): move ui and class fields under items

It was under groups by mistake.

J-Michalek commented:
Thanks for the fixes, seems I was pretty sloppy 🙄 I've implemented `item.ui` and `item.class` for `CommandPalette` and removed the tests for `item.ui` and `item.class`.

benjamincanac commented on docs/content/3.components/command-palette.md:
```diff
@@ -53,6 +53,8 @@ Each group contains an `items` array of objects that define the commands. Each i
 - [`slot?: string`{lang="ts-type"}](#with-custom-slot)
 - `onSelect?(e?: Event): void`{lang="ts-type"}
+- `ui?: { item?: ClassNameValue, itemLeadingIcon?: ClassNameValue, itemLeadingAvatarSize?: ClassNameValue, itemLeadingAvatar?: ClassNameValue, itemLeadingChipSize?: ClassNameValue, itemLeadingChip?: ClassNameValue, itemLabel?: ClassNameValue, itemLabelPrefix?: ClassNameValue, itemLabelBase?: ClassNameValue, itemLabelSuffix?: ClassNameValue, itemTrailing?: ClassNameValue, itemTrailingKbds?: ClassN ...
+- `class?: any`{lang="ts-type"}
```
```suggestion
- `class?: any`{lang="ts-type"}
- `ui?: { item?: ClassNameValue, itemLeadingIcon?: ClassNameValue, itemLeadingAvatarSize?: ClassNameValue, itemLeadingAvatar?: ClassNameValue, itemLeadingChipSize?: ClassNameValue, itemLeadingChip?: ClassNameValue, itemLabel?: ClassNameValue, itemLabelPrefix?: ClassNameValue, itemLabelBase?: ClassNameValue, itemLabelSuffix?: ClassNameValue, itemTrailing?: ClassNameValue, itemTrailingKbds?: ClassNameValue, itemTrailingKbdsSize?: ClassNameValue, itemTrailingHighlightedIcon?: ClassNameValue, itemTrailingIcon?: ClassNameValue,}`{lang="ts-type"}
```

Commit:
docs(command-palette): move ui under class for consistency

benjamincanac committed:
up

benjamincanac commented:
Not at all! You did an amazing job, thanks a lot 😊

benjamincanac commented:
@J-Michalek I've made a mistake regarding https://github.com/nuxt/ui/pull/4060#issuecomment-2868977199 and reverted it here: https://github.com/nuxt/ui/commit/0905b2b3d5e99ac78740a15d7a1afa1263ac7491.

My argument was that for consistency the `item.class` should pair with `ui.item` but as I was refactoring `@nuxt/ui-pro` to follow these changes, it became clear that the consistency is to target the same elements. For example, with this revert the `item.class` will be on `ui.link` but target the `LinkBase`, like on the DropdownMenu: https://github.com/nuxt/ui/blob/v3/src/runtime/components/DropdownMenuContent.vue#L175 and all other components.

Let me know if I forgot some other places.

J-Michalek commented:
Looks good, I checked my commits that made the original breaking changes and they only affected the `Breadcrumb` and `NavigationMenu` components which you both reverted correctly.
